src/Microsoft.TestPlatform.Common/Filtering/FilterExpression.cs

```diff
@@ -274,19 +303,49 @@ internal static FilterExpression Parse(string filterString, out FastFilter? fast
     internal bool Evaluate(Func<string, object?> propertyValueProvider)
     {
         ValidateArg.NotNull(propertyValueProvider, nameof(propertyValueProvider));
-        bool filterResult = false;
-        if (null != _condition)
-        {
-            filterResult = _condition.Evaluate(propertyValueProvider);
-        }
-        else
+
+        FilterExpression? current = this;
+        Stack<FilterExpression> filterStack = new();
+        Stack<bool> result = new();
+        do
         {
-            // & or | operator
-            bool leftResult = _left!.Evaluate(propertyValueProvider);
-            bool rightResult = _right!.Evaluate(propertyValueProvider);
-            filterResult = _areJoinedByAnd ? leftResult && rightResult : leftResult || rightResult;
-        }
-        return filterResult;
+            //Push root's right child and then root to stack then Set root as root's left child
+            while (current != null)
+            {
+                if (current._right != null)
+                {
+                    filterStack.Push(current._right);
+                }
+                filterStack.Push(current);
+                current = current._left;
+            }
+            //If the popped item has a right child and the right child is at top of stack, then remove the right child from stack, push the root back and set root as root's right child.
+            current = filterStack.Pop();
+            if (filterStack.Count > 0 && current._right == filterStack.First())
+            {
+                filterStack.Pop();
+                filterStack.Push(current);
+                current = current._right;
+            }
+            else
```

Comment:
This method is exactly the same as the one above except for the content of this if. I am wondering if it's worth having a private method for the code part and just passing a lambda for this part?

@nohwnd WDYT?

Reply:
Sounds good. It seemed like the stack managing logic could be nice to split from the actual filtering.